File: mathai/fraction.py
from .base import *
from .simplify import simplify
from .expand import expand

def fraction(expr):
    if expr is None:
        return None

    expr = simplify(expr)

    if expr.children == []:
        return expr

    children = [fraction(c) for c in expr.children]

    if expr.name == "f_add":
        terms = []

        for c in children:

            if c.name == "f_mul":
                num = []
                den = []
                for f in c.children:
                    if (
                        f.name == "f_pow"
                        and f.children[1].name.startswith("d_")
                        and int(f.children[1].name[2:]) < 0
                    ):
                        n = int(f.children[1].name[2:])
                        den.append(
                            f.children[0]
                            if n == -1
                            else TreeNode("f_pow", [f.children[0], tree_form(f"d_{-n}")])
                        )
                    else:
                        num.append(f)
                terms.append((num, den))

            elif (
                c.name == "f_pow"
                and c.children[1].name.startswith("d_")
                and int(c.children[1].name[2:]) < 0
            ):
                n = int(c.children[1].name[2:])
                terms.append(([], [
                    c.children[0]
                    if n == -1
                    else TreeNode("f_pow", [c.children[0], tree_form(f"d_{-n}")])
                ]))

            else:
                terms.append(([c], []))

        if not any(den for _, den in terms):
            return TreeNode("f_add", children)

        num_terms = []
        for i, (num_i, _) in enumerate(terms):
            acc = list(num_i)
            for j, (_, den_j) in enumerate(terms):
                if i != j:
                    acc += den_j
            if not acc:
                acc = [tree_form("d_1")]
            num_terms.append(
                acc[0] if len(acc) == 1 else TreeNode("f_mul", acc)
            )

        numerator = TreeNode("f_add", num_terms)

        den_all = []
        for _, den in terms:
            den_all += den

        denom = den_all[0] if len(den_all) == 1 else TreeNode("f_mul", den_all)
        denom = TreeNode("f_pow", [denom, tree_form("d_-1")])

        return simplify(
            TreeNode(
                "f_mul",
                [simplify(expand(numerator)), denom],
            )
        )

    return TreeNode(expr.name, children)

File: haskell/Fraction.hs
module Fraction (fraction) where

import Base
import Simplify (simplify)
import Expand (expand)
import Data.List (isPrefixOf)

fraction :: TreeNode -> TreeNode
fraction expr =
  let expr' = simplify expr
  in case children expr' of
       [] -> expr'
       _  ->
         let children' = map fraction (children expr')
         in case name expr' of
              "f_add" -> fractionAdd expr' children'
              _       -> TreeNode (name expr') children'


--------------------------------------------------
-- ADDITION CASE
--------------------------------------------------

fractionAdd :: TreeNode -> [TreeNode] -> TreeNode
fractionAdd _ children =
  let terms = map splitTerm children
  in if not (any (not . null . snd) terms)
        then TreeNode "f_add" children
        else buildFraction terms


--------------------------------------------------
-- SPLIT TERM INTO (NUMERATOR FACTORS, DENOMINATOR FACTORS)
--------------------------------------------------

splitTerm :: TreeNode -> ([TreeNode], [TreeNode])
splitTerm c =
  case name c of
    "f_mul" ->
      foldr splitFactor ([], []) (children c)

    "f_pow"
      | isNegativePower c ->
          let (base, n) = powInfo c
          in ([], [powDen base n])

    _ -> ([c], [])


splitFactor :: TreeNode -> ([TreeNode], [TreeNode]) -> ([TreeNode], [TreeNode])
splitFactor f (num, den)
  | isNegativePower f =
      let (base, n) = powInfo f
      in (num, powDen base n : den)
  | otherwise = (f : num, den)


--------------------------------------------------
-- NEGATIVE POWER HELPERS
--------------------------------------------------

isNegativePower :: TreeNode -> Bool
isNegativePower (TreeNode "f_pow" [_, e]) =
  case name e of
    s | "d_" `isPrefixOf` s -> read (drop 2 s) < 0
    _ -> False
isNegativePower _ = False


powInfo :: TreeNode -> (TreeNode, Int)
powInfo (TreeNode "f_pow" [b, e]) =
  (b, read (drop 2 (name e)))
powInfo _ = error "powInfo: not a power node"


powDen :: TreeNode -> Int -> TreeNode
powDen base n
  | n == -1   = base
  | otherwise = TreeNode "f_pow" [base, treeForm ("d_" ++ show (-n))]


--------------------------------------------------
-- BUILD FINAL FRACTION
--------------------------------------------------

buildFraction :: [([TreeNode], [TreeNode])] -> TreeNode
buildFraction terms =
  let numTerms =
        [ buildMul (num ++ concat [denj | (j, (_, denj)) <- zip [0..] terms, i /= j])
        | (i, (num, _)) <- zip [0..] terms
        ]

      numerator = TreeNode "f_add" numTerms

      denAll = concatMap snd terms
      denomMul =
        case denAll of
          [d] -> d
          _   -> TreeNode "f_mul" denAll

      denom = TreeNode "f_pow" [denomMul, treeForm "d_-1"]

  in simplify $
       TreeNode "f_mul"
         [ simplify (expand numerator "*")
         , denom
         ]


buildMul :: [TreeNode] -> TreeNode
buildMul []  = treeForm "d_1"
buildMul [x] = x
buildMul xs  = TreeNode "f_mul" xs
